extension/parquet/column_writer.cpp

```diff
@@ -1150,24 +1186,167 @@ class StringColumnWriter : public BasicColumnWriter {
+				}
+			}
+		} else {
+			// dictionary based page
+			for (idx_t r = chunk_start; r < chunk_end; r++) {
+				if (mask.RowIsValid(r)) {
+					auto value_index = page_state.dictionary.at(ptr[r].GetString());
+					if (!page_state.written_value) {
+						// first value
+						// write the bit-width as a one-byte entry
+						temp_writer.Write<uint8_t>(page_state.bit_width);
+						// now begin writing the actual value
+						page_state.encoder.BeginWrite(temp_writer, value_index);
+						page_state.written_value = true;
+					} else {
+						page_state.encoder.WriteValue(temp_writer, value_index);
+					}
+				}
 			}
 		}
 	}
 
-	idx_t GetRowSize(Vector &vector, idx_t index) override {
-		auto strings = FlatVector::GetData<string_t>(vector);
-		return strings[index].GetSize();
+	unique_ptr<ColumnWriterPageState> InitializePageState(BasicColumnWriterState &state_p) override {
+		auto &state = (StringColumnWriterState &)state_p;
+		return make_unique<StringWriterPageState>(state.key_bit_width, state.dictionary);
+	}
+
+	void FlushPageState(Serializer &temp_writer, ColumnWriterPageState *state_p) override {
+		auto &page_state = (StringWriterPageState &)*state_p;
+		if (page_state.bit_width != 0) {
+			if (!page_state.written_value) {
+				// all values are null
+				// just write the bit width
+				temp_writer.Write<uint8_t>(page_state.bit_width);
+				return;
+			}
+			page_state.encoder.FinishWrite(temp_writer);
+		}
+	}
+
+	duckdb_parquet::format::Encoding::type GetEncoding(BasicColumnWriterState &state_p) override {
+		auto &state = (StringColumnWriterState &)state_p;
+		return state.key_bit_width == 0 ? Encoding::PLAIN : Encoding::RLE_DICTIONARY;
+	}
+
+	void FlushDictionary(BasicColumnWriterState &state_p, ColumnWriterStatistics *stats_p) override {
+		auto &stats = (StringStatisticsState &)*stats_p;
+		auto &state = (StringColumnWriterState &)state_p;
+		if (state.key_bit_width != 0) {
+			// first we need to sort the values in index order
+			auto values = vector<string>(state.dictionary.size());
+			for (auto iter = state.dictionary.cbegin(); iter != state.dictionary.cend(); iter++) {
+				D_ASSERT(values[iter->second].empty());
+				values[iter->second] = iter->first;
+			}
+			// first write the contents of the dictionary page to a temporary buffer
+			auto temp_writer = make_unique<BufferedSerializer>();
+			for (idx_t r = 0; r < values.size(); r++) {
+				auto &value = values[r];
+				// update the statistics
+				stats.Update(value);
+				// write this string value to the dictionary
+				temp_writer->Write<uint32_t>(value.size());
+				temp_writer->WriteData((const_data_ptr_t)value.data(), value.size());
+			}
+			// flush the dictionary page and add it to the to-be-written pages
+			WriteDictionary(state, move(temp_writer), values.size());
+		}
+	}
+
+	idx_t GetRowSize(Vector &vector, idx_t index, BasicColumnWriterState &state_p) override {
+		auto &state = (StringColumnWriterState &)state_p;
+		if (state.key_bit_width == 0) {
```

Comment:
Perhaps this could be a function (e.g. `state.IsDictionaryEncoded()` or similar)?